lib/util/dependent-resource.js

```diff
@@ -0,0 +1,162 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+This file contains an abstract implementation for dependent metadata resources:
+key-value resources directly attached to a parent (e.g. tags, config variables).
+*/
+
+import * as Promise from 'bluebird';
+
+import {
+	findCallback,
+	isId,
+	isUnauthorizedResponse,
+	mergePineOptions,
+} from '../util';
+import { getUpsertHelper } from '../util/upsert';
+
+export function buildDependentResource(
+	{ pine },
+	{
+		resourceName, // e.g. device_tag
+		resourceKeyField, // e.g. tag_key
+		parentResourceName, // e.g. device
+		getResourceId, // e.g. getId(uuidOrId)
+	},
+) {
+	const upsert = getUpsertHelper({ pine });
+
+	const exports = {
+		getAll(options, callback) {
+			if (options == null) {
+				options = {};
+			}
+			callback = findCallback(arguments);
+
+			return pine
+				.get({
+					resource: resourceName,
+					options: mergePineOptions(
+						{ $orderby: `${resourceKeyField} asc` },
+						options,
+					),
+				})
+				.asCallback(callback);
+		},
+
+		getAllByParent(parentParam, options, callback) {
+			if (options == null) {
+				options = {};
+			}
+			callback = findCallback(arguments);
+
+			return getResourceId(parentParam)
+				.then(id =>
+					exports.getAll(
+						mergePineOptions(
+							{
+								$filter: { [parentResourceName]: id },
+								$orderby: `${resourceKeyField} asc`,
+							},
+							options,
+						),
+					),
+				)
+				.asCallback(callback);
+		},
+
+		get(parentParam, key, callback) {
+			callback = findCallback(arguments);
+
+			return getResourceId(parentParam)
+				.then(id =>
+					pine.get({
+						resource: resourceName,
+						options: {
+							$filter: {
+								[parentResourceName]: id,
+								[resourceKeyField]: key,
+							},
+						},
+					}),
+				)
+				.then(function(results) {
+					if (results[0]) {
+						return results[0].value;
+					}
+				})
+				.asCallback(callback);
+		},
+
+		set(parentParam, key, value, callback) {
+			return Promise.try(function() {
+				value = String(value);
+
+				// Trying to avoid an extra HTTP request
+				// when the provided parameter looks like an id.
+				// Note that this throws an exception for missing names/uuids,
+				// but not for missing ids
+				if (isId(parentParam)) {
+					return parentParam;
+				} else {
+					return getResourceId(parentParam);
+				}
+			})
+				.then(parentId =>
+					upsert(
+						{
+							resource: resourceName,
+							body: {
+								[parentResourceName]: parentId,
+								[resourceKeyField]: key,
+								value,
+							},
+						},
+						[parentResourceName, resourceKeyField],
+					).tapCatch(isUnauthorizedResponse, function() {
+						// On Pine 7, when the post throws a 401
```

Comment:
Is the comment still valid?

Reply:
I think so, if you try to link to a non-existent resource then I'd expect it to always be a 401 - you don't have access to that resource and we don't want to distinguish between it existing but you don't have access or it just not existing, as that would leak info that we don't want to leak

Reply:
@Page I would actually expect a 404, but I guess if we always return a 401 regardless if it exists or you don't have access, then it's fine.